cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java

```diff
@@ -200,21 +187,54 @@ private void closeDataset(DatasetCacheKey key, Dataset dataset) {
   }
 
   @Override
+  public void dismissDataset(Dataset dataset) {
+    Preconditions.checkNotNull(dataset);
+    // static datasets cannot be dismissed
+    if (staticDatasets.containsValue(dataset)) {
+      LOG.warn("Attempt to dismiss static dataset {} from dataset cache", dataset);
+      return;
+    }
+    if (txContext == null || !(dataset instanceof TransactionAware)) {
+      dismissSafely(dataset);
+    } else {
+      // it is a tx-aware: it may participate in a transaction, so mark it as to be dismissed after the tx:
+      // the transaction context will call dismissSafely() for this dataset when the tx is complete.
+      txContext.dismissAfterTx((TransactionAware) dataset);
+    }
+    // remove from activeTxAwares in any case - a dismissed dataset does not need to participate in external tx
+    // iterates over all datasets but we do not expect this map to become large
+    for (Map.Entry<DatasetCacheKey, TransactionAware> entry : activeTxAwares.entrySet()) {
+      if (dataset == entry.getValue()) {
+        activeTxAwares.remove(entry.getKey());
+        return;
+      }
+    }
+  }
+
+  /**
+   * Dismiss a dataset when it is known that no transaction is going on.
+   *
+   * @param dataset this is an Object because we need to pass in TransactionAware or Dataset
+   */
+  public void dismissSafely(Object dataset) {
+    // iterates over all datasets but we do not expect this map to become large
+    for (Map.Entry<DatasetCacheKey, Dataset> entry : datasetCache.asMap().entrySet()) {
+      if (dataset == entry.getValue()) {
+        datasetCache.invalidate(entry.getKey());
+        return;
+      }
+    }
+    // we can only hope that dataset.toString() is meaningful
+    LOG.warn("Attempt to dismiss dataset that was not acquired through this provider: {}", dataset);
+  }
+
+  @Override
   public TransactionContext newTransactionContext() {
     dismissTransactionContext();
-    txContext = new WeakReferencePromotingTransactionContext(txClient);
+    txContext = new DelayedDismissingTransactionContext(txClient);
     // make sure all in-progress participate in the transaction. But beware that weak refs may have expired
-    List<DatasetCacheKey> expiredTxAwares = new ArrayList<>();
-    for (Map.Entry<DatasetCacheKey, WeakReferenceDelegatingTransactionAware> entry : activeTxAwares.entrySet()) {
-      if (entry.getValue().isExpired()) {
-        expiredTxAwares.add(entry.getKey());
-      } else {
-        txContext.addTransactionAware(entry.getValue());
-      }
-    }
-    // remove the expired dynamic datasets
-    for (DatasetCacheKey key : expiredTxAwares) {
-      activeTxAwares.remove(key);
+    for (Map.Entry<DatasetCacheKey, TransactionAware> entry : activeTxAwares.entrySet()) {
+      txContext.addTransactionAware(entry.getValue());
```

Comment:
Yeah that ctor is now possible. The old code had to do extra work around this... now we can simplify.